compiler/AST/primitive.cpp

```diff
@@ -463,13 +468,66 @@ static QualifiedType
 returnInfoToBorrowed(CallExpr* call) {
   Type* t = call->get(1)->getValType();
 
-  if (UnmanagedClassType* mt = toUnmanagedClassType(t)) {
-    t = mt->getCanonicalClass();
+  ClassTypeDecorator decorator = CLASS_TYPE_BORROWED;
+  if (DecoratedClassType* dt = toDecoratedClassType(t)) {
+    t = dt->getCanonicalClass();
+    if (dt->isNilable())
+      decorator = CLASS_TYPE_BORROWED_NILABLE;
   }
+
+  if (decorator != CLASS_TYPE_BORROWED) { // leave borrow = canonical
+    if (AggregateType* at = toAggregateType(t)) {
+      if (isClass(at)) {
+        t = at->getDecoratedClass(decorator);
+      }
+    }
+  }
+
   // Canonical class type is borrow type
   return QualifiedType(t, QUAL_VAL);
 }
 
+static QualifiedType
+returnInfoToNilable(CallExpr* call) {
+  Type* t = call->get(1)->getValType();
+
+  ClassTypeDecorator decorator = CLASS_TYPE_BORROWED_NILABLE;
+  if (DecoratedClassType* dt = toDecoratedClassType(t)) {
+    t = dt->getCanonicalClass();
+    decorator = dt->getDecorator();
+    decorator = addNilableToDecorator(decorator);
+  }
+
+  if (AggregateType* at = toAggregateType(t)) {
+    if (isClass(at)) {
+      t = at->getDecoratedClass(decorator);
+    }
+  }
+
+  return QualifiedType(t, QUAL_VAL);
+}
+
+static QualifiedType
+returnInfoToNonNilable(CallExpr* call) {
+  Type* t = call->get(1)->getValType();
+
+  ClassTypeDecorator decorator = CLASS_TYPE_BORROWED;
+  if (DecoratedClassType* dt = toDecoratedClassType(t)) {
+    t = dt->getCanonicalClass();
+    decorator = dt->getDecorator();
+    decorator = removeNilableFromDecorator(decorator);
+  }
+
+  if (AggregateType* at = toAggregateType(t)) {
+    if (isClass(at)) {
+      if (decorator != CLASS_TYPE_BORROWED) // leave borrowed == canonical
+        t = at->getDecoratedClass(decorator);
```

Comment:
The above comment makes me think that `at->getDecoratedClass(CLASS_TYPE_BORROWED)` is different from `at`, which I think is not the case.

Please either adjust the comment or maybe just take out that conditional, because getDecoratedClass() will take care of it?